frontend/public/components/storage/attach-storage.tsx

```diff
@@ -40,306 +27,312 @@ const PVCDropdown: React.SFC<PVCDropdownProps> = props => {
+  const [error, setError] = React.useState('');
+  const [showCreatePVC, setShowCreatePVC] = React.useState('existing');
+  const [newPVCObj, setNewPVCObj] = React.useState(null);
 
-  componentDidMount() {
-    const { kindObj, name, namespace } = this.props;
-    const supportedKinds = [
-      'Deployment',
-      'DeploymentConfig',
-      'ReplicaSet',
-      'ReplicationController',
-      'StatefulSet',
-      'DaemonSet',
-    ];
-
-    if (!kindObj || !_.includes(supportedKinds, kindObj.kind)) {
-      this.setState({ error: 'Unsupported kind.' });
-      return;
-    }
+  const { kindObj, resourceName, namespace } = props;
+  const supportedKinds = [
+    'Deployment',
+    'DeploymentConfig',
+    'ReplicaSet',
+    'ReplicationController',
+    'StatefulSet',
+    'DaemonSet',
+  ];
 
+  React.useEffect(() => {
     // Get the current resource so we can add to its definition
-    k8sGet(kindObj, name, namespace).then(
-      resourceObj => {
-        this.setState({ resourceObj });
-      },
-      err => this.setState({ error: err.message })
-    );
+    k8sGet(kindObj, resourceName, namespace).then(setObj);
+  }, [kindObj, resourceName, namespace]);
+
+  if (!kindObj || !_.includes(supportedKinds, kindObj.kind)) {
+    setError('Unsupported kind.');
+    return;
   }
 
-  updateVolumeName = (claimName: string) => {
-    const { resourceObj } = this.state;
+  const updateVolumeName = (newClaimName: string) => {
     // Check if there is already a volume for this PVC.
-    const volumes = _.get(resourceObj, 'spec.template.spec.volumes');
+    const volumes = _.get(obj, 'spec.template.spec.volumes');
     const volume = _.find(volumes, {
       persistentVolumeClaim: {
-        claimName,
+        claimName: newClaimName,
       },
     }) as any;
 
-    const volumeName = volume ? volume.name : claimName;
-    const volumeAlreadyMounted = !!volume;
-    this.setState({ volumeName, volumeAlreadyMounted });
+    const newVolumeName = volume ? volume.name : newClaimName;
+    const newVolumeAlreadyMounted = !!volume;
+    setVolumeName(newVolumeName);
+    setVolumeAlreadyMounted(newVolumeAlreadyMounted);
+  };
+
+  const handleShowCreatePVCChange: React.ReactEventHandler<HTMLInputElement> = event => {
+    setShowCreatePVC(event.currentTarget.value);
+  };
+
+  const handleSelectContainers = () => {
+    setUseContainerSelector(!useContainerSelector);
+    setSelectedContainers([]);
+    setAllContainers(true);
   };
 
-  handleChange: React.ReactEventHandler<HTMLInputElement> = event => {
-    const { name, value } = event.currentTarget;
-    this.setState({ [name]: value } as any);
+  const handleContainerSelectionChange = (checked, event) => {
+    const checkedItems = [...selectedContainers];
+    checked ? checkedItems.push(event.currentTarget.id) : _.pull(checkedItems, event.currentTarget.id);
+    setSelectedContainers(checkedItems);
+    setAllContainers(checkedItems.length === 0);
```

Comment:
This doesn't seem right. We should show an error on save instead. (Looks like you're doing that below.)

Reply:
Can you elaborate? `AllContainers` is the flag that tells us to use all the containers when we are not using the container selector. It is a relic from the existing code.  Now that I have the `useContainerSelector` flag, I can get rid of it.

Reply:
I removed the state variable allContainers since it was repetitive.